app/controllers/spree/admin/payment_methods_controller.rb

```diff
@@ -191,6 +193,18 @@ def clear_preference_cache
           Rails.cache.delete(@payment_method.calculator.preference_cache_key(key))
         end
       end
+
+      def match_calculator_type
+        return if @payment_method.calculator.type == update_params['calculator_type']
+
+        new_calc = @object.build_calculator(type: update_params['calculator_type']&.constantize)
+        update_params['calculator_attributes']&.each do |k, v|
+          new_calc.method("#{k}=").call(v)
+        end
+        @payment_method = @object
+        update_params.delete('calculator_type')
+        update_params.delete('calculator_attributes')
+      end
```

Comment:
`±` Sorry I can't see why this was needed, would you be able to add a comment to the method, to explain?

Reply:
The method prevents ```ActiveModel::UnknownAttributeError``` from being raised when a user changes the calculator type and attempts to submit new calculator values while having unresolved errors elsewhere (such as invalid name or hub). After form submission, Rails would build a generic ```Spree::Calculator``` and tries to assign attributes like ```:preferred_flat_percent``` that don't exist on ```Spree::Calculator```, so it raises the error. 
My solution was to build and associate the correct calculator to the payment method object first, prior to calling ```@payment_method.update(update_params)```
Actually, I have found a better solution and the above method will be removed. I will still include a comment with the new code.

Reply:
Great, I found the new solution in aad3904  and it looks nice and simple 👍